pyvrp/cpp/search/primitives.h

```diff
@@ -0,0 +1,61 @@
+#ifndef PYVRP_SEARCH_PRIMITIVES_H
+#define PYVRP_SEARCH_PRIMITIVES_H
+
+#include "CostEvaluator.h"
+#include "Measure.h"
+#include "Route.h"
+
+// This file stores a few basic functions for (precisely) evaluating really
+// common moves. Those primitives may be useful implementing higher order
+// operators.
+namespace pyvrp::search
+{
+// Evaluates inserting U after V.
+/**
+ * Evaluates the delta cost of inserting U after V in V's route. The evaluation
+ * is exact.
```

Comment:
Question: regarding "the evaluation is exact", where in the codebase are the evaluations not exact?

Reply:
A lot of places shortcut when they know the delta cost will not become negative if more is evaluated. Some places even do this inexactly, like SWAP*.

These two don't, which is needed for e.g. rewriting SWAP* to use these: that operator evaluates multiple removes and inserts before deciding on the overall delta cost. It cannot do that if these 'building blocks' already shortcut.

Reply:
Ah okay, that makes more sense now!